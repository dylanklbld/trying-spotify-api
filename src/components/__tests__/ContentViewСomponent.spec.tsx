import { fireEvent, render, screen } from '@testing-library/react';

import ContentView小omponent from '../ContentView小omponent';
import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';

jest.mock("../Spinner", () => ({Spinner:(props) => <div data-testid="spinner" {...props} />}));
jest.mock("../Buttons/ScrollUpButton.tsx", () => (props) => <div data-testid="scroll-button" {...props} />);


const handleFetchMoreMock = jest.fn().mockName('handleFetchMore')
const scrollIntoViewMock = jest.fn().mockName('scrollIntoView');

const getItems = (amount = 10) => [...new Array(amount).keys()].map((k, i) => ({
    id: i,
    text: `data-${k}`
}))

type TestProps ={
    items: Array<any>,
    isLoading: boolean,
    canFetchMore: boolean
}

const renderComponent = ({items, isLoading = false, canFetchMore = false }:TestProps)=>{
    return render(<ContentView小omponent 
        items={items}
        isLoading={isLoading}
        canFetchMore={canFetchMore}
        handleFetchMore={handleFetchMoreMock}
        renderItem={(item:any)=><div key={item.id}>{item.text}</div>}
    />)
}

describe('ContentView小omponent', () => {
    it('should render simple case', () => {
        const items = getItems()
        const {container} = renderComponent({items})
        expect(container).toMatchSnapshot()
    })
    
    it('should have showMore button', () => {
        const items = getItems()
        const {container} = renderComponent({items, canFetchMore:true})
        screen.getByText("Show more")
        
        expect(container).toMatchSnapshot()
    })
    
    it('should call fetchMore', () => {
        const items = getItems()
        renderComponent({items, canFetchMore:true})
        fireEvent.click(screen.getByText("Show more"))
        
        expect(handleFetchMoreMock).toHaveBeenCalled()
    })
    
    it('should have loading status', async () => {
        const items = getItems()
        renderComponent({items, isLoading: true, canFetchMore:true})

        const showMoreButton = screen.getByText("Show more")
        expect(showMoreButton).toBeDisabled()
        
        const spinners = await screen.findAllByTestId("spinner")
       
        expect(spinners.length).toBe(2)

        fireEvent.click(showMoreButton)
        expect(handleFetchMoreMock).not.toHaveBeenCalled()
    })

    it('should call scroll into view', async () => {
        global.window = Object.create(window);
        window.HTMLElement.prototype.scrollIntoView = scrollIntoViewMock;       
        
        const items = getItems()
        renderComponent({items, canFetchMore:true})
        fireEvent.click(screen.getByTestId("scroll-button"))
        expect(scrollIntoViewMock).toHaveBeenCalled()
    })
  })