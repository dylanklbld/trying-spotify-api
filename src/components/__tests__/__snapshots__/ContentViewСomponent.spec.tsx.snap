// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentViewСomponent should have showMore button 1`] = `
<div>
  <div
    id="scroll-top-anchor"
  />
  <div
    class="sc-bdfBwQ dKNVSb"
  >
    <div>
      data-0
    </div>
    <div>
      data-1
    </div>
    <div>
      data-2
    </div>
    <div>
      data-3
    </div>
    <div>
      data-4
    </div>
    <div>
      data-5
    </div>
    <div>
      data-6
    </div>
    <div>
      data-7
    </div>
    <div>
      data-8
    </div>
    <div>
      data-9
    </div>
  </div>
  <div
    data-testid="scroll-button"
  />
  <div
    class="sc-gsTCUz eBajvn"
  >
    <button
      class="sc-dlfnbm elvkCX"
    >
      Show more
    </button>
  </div>
</div>
`;

exports[`ContentViewСomponent should render simple case 1`] = `
<div>
  <div
    id="scroll-top-anchor"
  />
  <div
    class="sc-bdfBwQ dKNVSb"
  >
    <div>
      data-0
    </div>
    <div>
      data-1
    </div>
    <div>
      data-2
    </div>
    <div>
      data-3
    </div>
    <div>
      data-4
    </div>
    <div>
      data-5
    </div>
    <div>
      data-6
    </div>
    <div>
      data-7
    </div>
    <div>
      data-8
    </div>
    <div>
      data-9
    </div>
  </div>
  <div
    data-testid="scroll-button"
  />
  <div
    class="sc-gsTCUz eBajvn"
  />
</div>
`;
